### ISSUE TITLE
ENA consensus submissions

### ISSUE THREAD

**Issue by SamStudio8** [COG-UK/dipi-group#11]
* [x] `institute.ena_assembly_opted`
* [x] How to get process information in scope (`process.bioinformatics.pipeline.pipe`) for `ocarina-get-ena-assembly.py`
* [x] How to get `ENA-SAMPLE` and `ENA-RUN` accessions in scope for "
* [x] How to get metrics in scope for "
* [x] Add v2 credit system to response
* [x] Script to fetch metadata via ocarina https://github.com/SamStudio8/coguk-push/commit/a2f9ef23ab64a44e012f16d6a3ce5965d1c5270a, https://github.com/SamStudio8/coguk-push/commit/ac3c022673734a5b78848b1db327eea086b42fea
* [x] Script to munge metadata to webin-cli calls
* [x] Nextflow pipeline for ENA-A (https://github.com/SamStudio8/elan-ena-nextflow)
* [x] Check accession receipt parsing
* [x] Script to load accessions into Majora (https://github.com/SamStudio8/coguk-push/commit/14a3dc3a62b4153ed55c02637f0f1fabf9a4fa4f)
* [x] Send UoB data
* [x] Integrate bioinformatics information to the uploader (https://github.com/SamStudio8/majora/issues/45)
* [x] ~Allow stomp-free updating of said new uploader fields~
* [x] Add opt-in to institute form
* [x] ~Future: Automatic mapping of `ERZ` accessions to real INDSC identifiers~

**Comment by SamStudio8** [comment]
We can get the requisite metrics and process information in scope through the `bind_to_metadata` script but we really ought to be thinking more outside the box on this. Might be time to break out the `v3` API and register an MDV.

**Comment by SamStudio8** [comment]
First single-sequence submission uploaded today through `webin-cli`, should be public tomorrow.

**Comment by SamStudio8** [comment]
https://github.com/SamStudio8/majora/commit/0740b4bd56af8c4383c2b98fb348718cfd5d8dc4 adds a new task `task_api_get_pags_to_publish`. Currently this can be executed by overriding the usual get pag interface with a new `mode` flag. This new function removes the need for serializers (serpy or DRF) entirely and is significant faster than the dynamic v3 API. The drawback is this adds another highly-specific endpoint to Majora which was not the intended design. The most important thing is it works for now and does at least allow us to merge the two ENA submission pipelines (as I went through the pain of ensuring the read-level information needed for ENA was included in this new interface); which means we can finally deprecate the horrible CSV binding step in the current ENA pipeline - fully decoupling the outbound pipelines from the ELAN1 metadata table.

**Comment by SamStudio8** [comment]
Credit codes implemented https://github.com/SamStudio8/majora/commit/784b3dd92a000e23a47a0d130d1745cfbb467494
I've additionally cleaned up the `/r/n` garbage people have managed to put in those fields.

**Comment by SamStudio8** [comment]
Significant work has gone into this over at https://github.com/SamStudio8/majora/issues/45. Majora, Ocarina, Elan and the metadata uploader have been updated to accommodate setting and updating of `pipe_name` and `pipe_version`.

**Comment by SamStudio8** [comment]
https://github.com/SamStudio8/coguk-push/commit/a2f9ef23ab64a44e012f16d6a3ce5965d1c5270a adds a script to fetch ENA-ASSEMBLY submission metadata with Ocarina. https://github.com/SamStudio8/coguk-push/commit/ac3c022673734a5b78848b1db327eea086b42fea adds a script to check that downloaded PAGs are eligible for submission (requirements: non-empty, non-default `assembler`, and valid ENA sample and ENA run accessions).

**Comment by SamStudio8** [comment]
ENA are chasing up an issue their side that has prevented the correct processing of my test submission, but otherwise all looks fine. We've been requested to maintain a link to the GISAID accession at the assembly level, however it is not possible to provide user-defined fields in the `webin-cli` manifest, so we cannot action that request. We've additionally been asked to send `SUBMISSION_TOOL` and `SUBMISSION_TOOL_VERSION` to indicate the name and version of the `webin-cli` wrapper.

* [x] Add GISAID strain ID (wontfix)
* [x] Add `SUBMISSION_TOOL` and `SUBMISSION_TOOL_VERSION` to manifest

**Comment by SamStudio8** [comment]
I don't think this behaviour really needs to be wrapped up in `pyena` because the way we fetch metadata from Majora is rather specific and won't generalise to be that useful. I'll instead add it to a new dir in the Elan nextflow repository. There's no need to roll some python just for generating the `chr_list` and `manifest` so we can just do it all in Nextflow.

**Comment by SamStudio8** [comment]
Making good progress on this. I've started the new Nextflow script (https://github.com/SamStudio8/elan-nextflow/blob/a214282daf47ed4dd8d2189b742979010e12ac4a/ena-assembly/ena-a.nf). We can generate the manifests and validate them with `webin-cli`. Little more cleaning up and I think we'll be ready to go.

```[0f/aa848d] process > prep_fasta        [100%] 15 of 15 ✔
[9a/274560] process > generate_chrlist  [100%] 15 of 15 ✔
[c0/09ad68] process > generate_manifest [100%] 15 of 15 ✔
[d3/e094b3] process > webin_validate    [100%] 15 of 15 ✔
```

**Comment by SamStudio8** [comment]
This stuff seemed pretty useful, so I've pulled it fron `elan-nextflow` into a [new `elan-ena-nextflow` repository](https://github.com/SamStudio8/elan-ena-nextflow) that can be called by anyone using the Nextflow pipeline sharing mechanism.

**Comment by SamStudio8** [comment]
Updated `elan-ena-nextflow`, it is now possible to successfully send test submissions, and parse the XML receipts into a new outfile:
```
nextflow run samstudio8/elan-ena-nextflow -r stable --study $WEBIN_STUDY --webin_jar webin-cli-3.4.0.jar --manifest ena-assembly-erz-test.csv --ascp --test -resume --out test.accessions.txt
[...]
[61/9ac4c1] process > prep_fasta        [100%] 15 of 15 ✔
[60/e95d0b] process > generate_chrlist  [100%] 15 of 15 ✔
[6d/00e2c3] process > generate_manifest [100%] 15 of 15 ✔
[1e/ec8447] process > webin_validate    [100%] 15 of 15 ✔
[03/a42c27] process > webin_submit      [100%] 15 of 15 ✔
[e7/b9d675] process > receipt_parser    [100%] 15 of 15 ✔
```

With that, we're basically in position to try real submissions.

* [x] Check internally what note to append to `DESCRIPTION` for QC/metadata.
* [x] Ask NR whether spaces require escaping in `manifest` (nope)
* [x] Ask NR whether newlines are allowed in `DESCRIPTION` (yes)

**Comment by SamStudio8** [comment]
Need some internal discussion to decide what extra information (particularly on QC) should be provided in the `DESCRIPTION` tag before we start uploading the sequences to ENA. I'm also checking why the `ASSEMBLYNAME` does not appear in genbank.

* [x] `ASSEMBLYNAME` not in genbank record - this is normal according to my contact at EBI

**Comment by SamStudio8** [comment]
DWG hasn't had a chance to meet yet. I've set up a holding page for a "[COG-UK data portal](https://github.com/COG-UK/data)" and will use the following text for an initial small-scale submission of `BIRM` sequences:

```
Sample metadata and QC flags may have been updated since deposition in public databases.
COG-UK recommends users refer to data.covid19.climb.ac.uk for latest metadata and QC tables
before conducting analysis.
```

**Comment by SamStudio8** [comment]
Overcoming another Groovy related confusion to fill in `DESCRIPTION` automatically; https://twitter.com/samstudio8/status/1364248843127304195. 

Update: https://github.com/SamStudio8/elan-ena-nextflow/commit/044d0fa8d0afe7fb9f35e0087a98d9399fa352af - `DESCRIPTION` can now be filled in all-smart-like.

**Comment by SamStudio8** [comment]
https://github.com/SamStudio8/coguk-push/commit/14a3dc3a62b4153ed55c02637f0f1fabf9a4fa4f adds a script to read accessions from `elan-ena-nextflow` into Majora. I've a test script that uses the `ocarina` fetching endpoint and`elan-ena-nextflow` to mint a real accession over at ENA, I've sent one single sequence and will wait for feedback from NR before submitting any more.

**Comment by SamStudio8** [comment]
DWG have met and decided the following:

* [x] BAM passing basic QC should be uploaded
* [ ] ~Documentation for `bioinfo_pipe_*` fields needs to be written~
* [x] Add best QC test to `DESCRIPTION` (anything else?)

Additionally;
* [x] Unify `DESCRIPTION` and any other applicable fields with WSI
* [x] Check last submission with NR
* [ ] System to scrape minted INSDC accessions from ENA

**Comment by SamStudio8** [comment]
NR has got back to me and said our recent submission has failed to generate properly due to the formatting of the `AUTHORS` line. This is a little frustrating as it passes validation in `webin-cli`! However, NR is checking what our options are with the database team. I'm assuming this is causing trouble converting the submission to an INSDC record which has very specific formatting.

**Comment by SamStudio8** [comment]
`AUTHORS` issue resolved last week by ENA. For the COG Webin account specifically, the authorship information will be automatically moved to a different field in the INSDC records to allow for authorship information to be included in the record, without requiring strict reformatting. Remaining is a small internal discussion on the format and contents of the `DESCRIPTION` field.

I have handled processing of the Basic QC BAM backlog.

**Comment by SamStudio8** [comment]
RD and I have settled on a `DESCRIPTION` scheme and RD has sent a new test record to ENA. Once this comes back I'll begin a small scale test and submit the Birmingham samples.

**Comment by SamStudio8** [comment]
RD test looks good with new agreed format. I need to add `ADDRESS`, get the QC data in scope of `DESCRIPTION` and then try to send off the UoB samples.

* [x] `ADDRESS`
* [x] New `QC` fields in `DESCRIPTION`
* [x] Upload Birmingham

**Comment by SamStudio8** [comment]
Need the Majora `get pag` endpoint to be able to reply with QC results

**Comment by SamStudio8** [comment]
https://github.com/SamStudio8/majora/commit/b8b3d0a16263180f82ba59f200d27fdee3b648e5 adds QC information to `get pag` API reply, https://github.com/SamStudio8/ocarina/commit/49c54dd514531e14148698ac40f1a32216ac419b allows Ocarina (v0.38.2) to use the `qc_reports` reply to pass through `PASS` and `FAIL` to the ENA-A CSV. https://github.com/SamStudio8/coguk-push/commit/80d19d52bc92c2c9c8f6e4935f0e63014422464d adds the handling for these new fields in the ENA-A prep script.

https://github.com/SamStudio8/coguk-push/commit/c8776909350de3a72fb9de1712d87a9ccb0ac1f2 Script to run submissions
https://github.com/SamStudio8/coguk-push/commit/e948083d3e626ad5c6e83213de0763187906700f Script to announce submissions

**Comment by SamStudio8** [comment]
Tested this end-to-end, a new Birmingham sample appears to have gone through in one piece! :tada: 


![Screenshot from 2021-03-30 15-49-48](https://user-images.githubusercontent.com/1105722/113009213-a9236a80-916f-11eb-878f-35458cd32bc4.png)

I'll wait for this to appear on ENA side and check for any final trouble with the `AUTHORS` and `ADDRESS`, as well as check the `DESCRIPTION` is not mangled.

**Comment by SamStudio8** [comment]
Last I heard the incoming data looked good. We've sent a few small scale tests of Birmingham data. Before the Easter break I was seeing a high rate of API call fails to ENA but I'm hoping that was just temporary. I'll proceed with larger incremental testing and then we can onboard a few sites.

**Comment by SamStudio8** [comment]
Tried a larger submission today with great success. Only 4 samples failed "because of a system [ENA-side] error". Interestingly the error message makes reference to FTP which makes me wonder if webin-cli is silently dropping to FTP because ascp is not on the PATH (which it should be).

* [ ] Check ascp is in PATH scope for ENA-A submissions

**Comment by SamStudio8** [comment]
Met with the data working group and decided:
* [x] Turn on opt-ins
* [x] Crowdsource the values for `bioinfo_pipe_name` and `bioinfo_pipe_version` to try and assist backfill

Presumably we'll need to provide some docs on this (someone could port the draft gdoc to the official docs site)

**Comment by SamStudio8** [comment]
Question from external data consumer about the `TITLE` field in our new Genbank records and whether we can set it to something other than "Direct submission". I'd previously asked NR at ENA about this but followed up for a definite answer this week. NR confirms the title cannot be populated from the ENA flatfile, but we'll make it clear to users on the data website that the `COG_ACCESSION` (that would be the likely value for `TITLE` if we could provide one) is provided in the `COMMENT` block.

**Comment by SamStudio8** [comment]
We need to issue the guidance on `bioinfo_pipe_name`, `bioinfo_pipe_version` and the process before we can turn on opt-ins.

**Comment by SamStudio8** [comment]
I caved and did it myself. https://github.com/COG-UK/docs/commit/f68a886613b2208ba90c106ea58be6930e0df9de ports the previous work-in-progress guidance to the official docs site and will maintain the table of recommendations. https://github.com/SamStudio8/majora/commit/c4e2b717ab0df58e069bfa7040874fc84294d013 enables opt-ins on the Majora institute account system.

I'll deploy the patch after the morning rush and contact a site to beta test.

* [x] Enrol a site for beta test
* [x] Submit

**Comment by SamStudio8** [comment]
Largest test submission yet! I'll ask NR to keep an eye on these this week and then we can open the proverbial floodgates and finally close this issue.

**Comment by SamStudio8** [comment]
Feature announced to `#metadata`, thrilled to finally close this one. This has been a huge amount of work covering teams at CLIMB, WSI, ENA and the INSDC. We've even produced a fancy pipeline (https://github.com/SamStudio8/elan-ena-nextflow) for others to use as a result :tada:
